hypershift-operator/controllers/controlplaneautoscaler/controller.go

```diff
@@ -0,0 +1,349 @@
+package controlplaneautoscaler
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sync"
+
+	"github.com/go-logr/logr"
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	hyperv1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	schedulingv1alpha1 "github.com/openshift/hypershift/api/scheduling/v1alpha1"
+	"github.com/openshift/hypershift/hypershift-operator/controllers/manifests"
+	controlplaneautoscalermanifests "github.com/openshift/hypershift/hypershift-operator/controllers/manifests/controlplaneautoscaler"
+	"github.com/openshift/hypershift/support/util"
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/types"
+	vpaautoscalingv1 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/utils/ptr"
+
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	ControllerName = "ControlPlaneAutoscaler"
+
+	// memorySizeFactor is used to determine whether the VPA recommendation fits within a cluster size.
+	// The HostedCluster will be sized to the smallest size for which capacity * memorySizeFactor >= recommendation.
+	memorySizeFactor = 0.65
+
+	// hosted cluster namespace annotation on VPA
+	hcNamespaceAnnotation = "hypershift.openshift.io/cluster-namespace"
+
+	// hosted cluster name annotation on VPA
+	hcNameAnnotation = "hypershift.openshift.io/cluster-name"
+)
+
+type ControlPlaneAutoscalerController struct {
+	client.Client
+	sizeCache machineSizes
+}
+
+type machineSizes struct {
+	cscGeneration int64
+	sizes         map[string]machineResources
+	m             sync.Mutex
+}
+
+type machineResources struct {
+	Memory resource.Quantity
+	CPU    resource.Quantity
+}
+
+func (s *machineSizes) update(csc *schedulingv1alpha1.ClusterSizingConfiguration, listMachineSets func() (*machinev1beta1.MachineSetList, error), log logr.Logger) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	if csc.Generation == s.cscGeneration {
+		// machine sizes are up to date
+		return nil
+	}
+
+	log.Info("Updating machine size cache")
+
+	machineSetList, err := listMachineSets()
```

Comment:
Is there a precedent for the HC to fetch management cluster machines?
would this need rbac adjustment?
Is there a reason why this info can't be made available within the ClusterSizingConfiguration API?

Reply:
I did consider that, but in addition to requiring an API change, it seemed that we would be asking you to provide redundant information. Right now the contract is that fleet manager produces machinesets that are labeled with corresponding size labels. They decide the instance type that goes into those machinesets when they create them. This would require that they now keep the clustersizingconfig in sync with whatever machine size they decide to use for each size label.